### DISCUSSION TITLE
Cookie Banner customization / Privacy Policy visibility

### DISCUSSION THREAD

**Discussion by declann** [quarto-dev/quarto-cli#5958] [Feature Requests]
### Description

Hi,

The default cookie banner text is:

`type: express`:
![image](https://github.com/quarto-dev/quarto-cli/assets/353214/891714dd-2246-474c-961a-f676e4aff0d3)

`type: implied`:
![image](https://github.com/quarto-dev/quarto-cli/assets/353214/27d740d6-76a2-41b0-9c8f-7dfb629d00ad)

I think it would be useful to be able to customise this, e.g. targeted ads aren't always relevant. Is this feasible?

Also or alternatively, whenever `policy-url` is set I think it'd be useful to include a link to the privacy policy page in the banner. Currently, that only appears under "Change My Preferences" and then "More information" - a little hidden from the user perspective?

![image](https://github.com/quarto-dev/quarto-cli/assets/353214/8d4cc79c-91f0-4b80-9fe6-6a43a7752bd0)

Reference: Quarto docs on cookie banner: https://quarto.org/docs/websites/website-tools.html#cookie-consent

Cheers,

Declan

**Comment by mcanouil** [comment]
The banner is made from <https://www.cookieconsent.com/> and not made by Quarto.
So you can customise and use your own.

**Reply by declann** [reply]
Thanks!

I was uncertain about a few things about this code: I could not find a license, => interpret if I am actually entitled to change/use it, and also source, because I could only find a minified file without source and that's also worrying for security, licensing, etc.

So I contacted TermsFeed who make this and the response I got indicates that the issues I mentioned (including better control over text) will be addressed in a new version and a new move to release on GitHub, which is promising but also not ideal in the now.

I was clearly interested in alternatives anyways (and in learning more about how much control I have over Quarto output) so I moved [my site](https://models-on-a-plane.pages.dev/) to [an alternative](https://github.com/orestbida/cookieconsent).

The changes are here: https://gitlab.com/declann/models-on-a-plane/-/blob/main/_quarto.yml and here: https://gitlab.com/declann/models-on-a-plane/-/blob/main/cookieconsent-init.js , though I haven't brought back analytics scripts conditional on preferences yet.

I can specify my own text in that init file, so it suits those needs. OTOH it isn't quite as encapsulated as the TemsFeed one Quarto uses by default (though it can _be_ encapsulated).

Since Quarto includes TermsFeed cookie banner now, the project perhaps might be interested in **analysing/tracking and/or mitigating** issues from the TermsFeed specifics I mentioned above, in any case.

**Reply by termsfeed** [reply]
Thanks for the feedback here! Following up here from our email discussion.

A link to the Privacy Policy inside the notice banner can be forced by adding `notice_banner_insert_legal_urls` to `true` in the configuration code.

However, the "targeted ads" part is pending from our end. We're working on a new version that should address this limitation in the notice banner text and to allow for more configuration over the text while keeping the 10+ language translations in sync. (For comments history, current version is 4.1.0)

The licensing and documentation will be provided on our Github repo soon as well.

**Reply by mcanouil** [reply]
FYI @dragonstyle

**Reply by mcanouil** [reply]
FYI:

- #433

**Comment by greg-dubrow** [comment]
Hi - Someone just opened an issue on my blog repo that I'm not GDPR compliant because I had the Ok and Change My Preferences 
<img width="1426" alt="Screenshot 2024-02-19 at 13 26 29" src="https://github.com/quarto-dev/quarto-cli/assets/7495863/b527c470-640e-4ada-936e-a1287c205773">

banner, and not the Agree - Decline - Change - Preferences. I've changed the quarto.yml to:
`  cookie-consent: true
      type: express`
But that hasn't changed the banner. Is there something I'm missing? I do have a google analytics enabled in the .yml could that be it?

I'm not expecting a code review or debugging, just wondering if anyone else has had this same issue. I thought the way I now have the .yml would be compliant. 

I like Derek's approach, just not sure how exactly to implement it without breaking things.

**Reply by declann** [reply]
Whos Derek :smile: 

Here is the repo for calculang.dev which is a Quarto site but using this cookie consent:
https://www.github.com/orestbida/cookieconsent

repo: https://gitlab.com/declann/calculang.dev

If you search `_quarto.yml` for cookie you'll see my cookie and opt-in analytics config.

You also need the `resources/cookieconsent-2.9.1` folder (not sure if there is something more recent) and also [cookieconsent-init.js](https://gitlab.com/declann/calculang.dev/-/blob/main/cookieconsent-init.js?ref_type=heads) (I modified this file so you'll want to review and it also includes e.g. a link to privacy policy so you might want to update that)

IANAL

That's it off the top of my head but you can reach out if you have Qs about this implementation.

I'm out of date on the Quarto cookie config so can't help there.

Regards
Declan

**Reply by greg-dubrow** [reply]
Doh, yes...Declan...sorry about that :)

I did see your yaml, the .js file and the folder. It's just a matter of adding them in appropriate places and modifying what you noted in this reply? Looking at the calculanf documentation they have updated it, so I'll grab that. Thanks!

**Reply by declann** [reply]
I think I noted it all for 2.9 version. The cookieconsent code is very vanilla and I just use Quarto features in `_quarto.yml` to get it included in the body and CSS in the head; then do ordinary cookieconsent config through the `-init.js` file.

In my config I also include analytics scripts which are opt-in so look like this:
`<script defer type="text/plain" data-cookiecategory="analytics" ...`
and in the `-init` file I created a toggle block for analytics.

Since 3.0 is out I would probably try it. I expect it to work in a similar way but would need to make similar config and test all the pieces.

I don't notice anything very different to get it working; in the Playground the script that corresponds to my `-init` is `-config` - that's all; but I haven't looked closely.

**Reply by termsfeed** [reply]
If you're using the Free Cookie Consent from TermsFeed, you need to update the config code with:

> consent_type: "express"

And make sure you're using the latest version: 4.1.0

**Reply by declann** [reply]
Hi, is there any update on clarity Re licensing and on providing source code (not minified) for TermsFeed cookie consent?

**Reply by termsfeed** [reply]
For the Free Cookie Consent? Yes, we're working on releasing the unminified versions for all versions in the near future. We're about to release 4.2.0 in the next few weeks and then release the unminified versions. I can notify via email if preferred.

**Reply by declann** [reply]
Thanks for the update! If you update here that would be great!

**Comment by NicolasBoumal** [comment]
A work-around to personalize the text that appears in the cookie consent banner is as follows. It uses JavaScript to replace the text in that banner with your own choice once the page is fully loaded. Not ideal, but it makes it possible to remove the mention of "targeted ads" for a website that has no ads.

_quarto.yml:
```
include-in-header:
  file: header.html
```

header.html:
```
<script>
  window.addEventListener('load', function() {
    document.getElementById("cc-nb-title").innerHTML = "My cookie banner title";
    document.getElementById("cc-nb-text").innerHTML = "My cookie banner text";
  });
</script>
```
